Sanitize and validate dashboards

IreneMLC opened the pull request:
### Relevant information

Add script to sanitize dashboard:
 - Anonymize account ID
 - Remove linkedEntityGuids
 - Remove permissions

Add JSON schema for the new dashboard format and validate dashboards. 

~Add pre-commit hook (sanitizes dashboards) and pre-push hook (validates definitions).~
Add a dockerized sanitizer and validator.

naxhh commented on validator/.husky/.gitignore:
```diff
@@ -0,0 +1 @@
+_
```
what magic is this one? :D

IreneMLC replied:
Husky added this on its own 😅 so I didn't touch it

naxhh commented on validator/.husky/pre-push:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+. "$(dirname "$0")/_/husky.sh"
+
+npm --prefix validator install && npm --prefix validator run check
```
can we don't do this one?

i'm very against not allowing to push because validations fail. There sometimes reasons to do that and this makes the experience a bit worst imho

IreneMLC replied:
Oki, I'll remove it (I think that people won't run them otherwise, but the PR will fail, so it can be fixed here).

naxhh reviewed (approved):
Nice!

Should we note in the readme that now you MUST have some version of node/npm installed to contribute?

Since the precommit will fail otherwise

srvaroa commented:
I would try to avoid forcing people to install stuff locally and instead provide a docker image to run them.

IreneMLC commented:
So, should I remove the hooks completely? (and we ask people to run the sanitization themselves?)

IreneMLC commented:
I removed the hooks so that people are not forced to install anything, and added the dockerized sanitizer (+ full validator), also updated the README, wdyt?

IreneMLC commented on validator/schemas/dashboard-schema-v1.json:
```diff
@@ -0,0 +1,229 @@
+                          "title": "The title of the widget"
+                        },
+                        "rawConfiguration": {
+                          "$id": "#/properties/pages/items/anyOf/0/properties/widgets/items/anyOf/0/properties/rawConfiguration",
+                          "type": "object",
+                          "title": "The configuration of the widget",
+                          "required": [
+                                "nrqlQueries"
+                              ],
+                          "not": { "required": [ "metrics" ] },
```
We've confirmed that it's not possible to filter metric-format widgets by entity guid, which we need for the explorer, so we should not allow dashboards that contain widgets in this format. The query builder now generates NRQL-format metrics so this should only show up if people export old dashboards and we can help them fix it.

cianBuckley commented on validator/README.md:
```diff
@@ -26,7 +42,8 @@ You can also run each of the validation tools independently:
 npm  --prefix validator run validate-schemas
-npm  --prefix validator run validate-uniqueness
+npm  --prefix validator run validate-rules
+npm  --prefix validator run validate-folders
```
worth adding `sanitize-dashboards` here?

IreneMLC replied:
Probably yes, I haven't mentioned it on any readme, thanks!

cianBuckley commented on validator/tools/sanitize_dashboards.js:
```diff
@@ -0,0 +1,42 @@
+(async () => {
+    const folderDefinitions = await readdir(DEFINITIONS_DIR, {withFileTypes: true})
+    for (let folderDefinition of folderDefinitions){
+        if(folderDefinition.isDirectory()){
```
worth raising errors on the negative of the folder and file branches?

IreneMLC replied:
If we have an unexpected file here you mean? 🤔 there's another script that validates folders so here I added this just in case

cianBuckley commented on validator/tools/sanitize_dashboards.js:
```diff
@@ -0,0 +1,42 @@
+                            .replace(/\"accountId\"\s*:\s*\d+\s*/g , '"accountId": 0') // Anonymize account ID
+                            .replace(/^.+\"linkedEntityGuids\".+\n?/mg , '') // Remove linkedEntityGuids
+                            .replace(/^.+\"permissions\".+\n?/mg , '') // Remove permissions
+                            .replace(/\,(?!\s*?[\{\[\"\'\w])/g, ''); // Remove trailing commas
```
You must be getting handy at the auld regex at this stage :D

IreneMLC replied:
😆 😭

apoloa commented on validator/tools/sanitize_dashboards.js:
```diff
@@ -0,0 +1,42 @@
+    for (let folderDefinition of folderDefinitions){
+        if(folderDefinition.isDirectory()){
+            const files = await readdir(path.resolve(DEFINITIONS_DIR, folderDefinition.name), {withFileTypes: true})
+            for(let file of  files) {
```
```suggestion
            for(let file of files) {
```

apoloa commented on validator/tools/sanitize_dashboards.js:
```diff
@@ -0,0 +1,42 @@
+                            process.exit(1)
+                        }
+
+                        let newFileContent = fileContent
```
Should be nice to have some tests to avoid problems in the future. 😹
